pkg/tfbridge/validate_input_types.go

```diff
@@ -0,0 +1,471 @@
+package tfbridge
+
+import (
+	"fmt"
+	"regexp"
+	"slices"
+	"strings"
+
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+)
+
+// pathBuilder is a helper for building property paths for
+// better error messages. As you recurse through objects and arrays
+// you can build up a path to the property that failed
+type pathBuilder struct {
+	paths  []string
+	isList bool
+}
+
+// add adds a property path to the path builder
+func (c *pathBuilder) add(path string) {
+	if !c.isList {
+		c.paths = append(c.paths, path)
+	}
+	c.isList = false
+}
+
+// addListIndex adds a list index to the path builder, but does not mutate the
+// original path builder. This allows for subsequent list items to have a
+// distinct path builder
+func (c *pathBuilder) addListIndex(idx int) pathBuilder {
+	cc := *c
+	cc.paths = append(c.paths, fmt.Sprintf("%d", idx))
+	cc.isList = true
+	return cc
+}
+
+// toPath renders the property path as a string
+func (c *pathBuilder) toPath() string {
+	return strings.Join(c.paths, ".")
+}
```

Comment:
Can we use [`resource.PropertyPath`](https://github.com/pulumi/pulumi/blob/28f098a660aa698c83a77058da28f7867707d0a0/sdk/go/common/resource/properties_path.go#L14-L16) here?

Reply:
It doesn't look like `PropertyPath` includes any utilities for building a path, it looks like it can create a `PropertyPath` from a string and can edit `PropertyValue`s.

Reply:
`PropertyPath` is defined as:

```go
type PropertyPath []interface{}
```

That means that `func (c *pathBuilder) add(path string)` is equivalent to `c = append(c, path)` and `func (c *pathBuilder) addListIndex(idx int)` is equivalent to `c = append(c, idx)`.

Reply:
Ok I think I was able to refactor it to work with `PropertyPath`!